### PR TITLE
fix #1967 MPP-1639: Add DomainAddress.used_on field

### LINKED ISSUE #1967: Add DomainAddress.used_on field
This is a counterpart request for work that has already been done on the `RelayAddress` object side: https://github.com/mozilla/fx-private-relay/pull/1713

See https://mozilla-hub.atlassian.net/browse/MPP-1639 for more info.

### PR DESCRIPTION
**Pull request by groovecoder**
This PR fixes #1967.

How to test:
* Create a domain address
* Go to http://127.0.0.1:8000/api/v1/domainaddresses/
  * [ ] Verify you see a new `used_on` field for the domain address

- [ ] I've added a unit test to test for potential regressions of this bug.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

### CODE REVIEW

File: emails/migrations/0042_domainaddress_used_on.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Generated by Django 3.2.13 on 2022-05-19 14:59
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('emails', '0041_relayaddress_used_on'),
```
Comment (jwhitlock) [inline emails/migrations/0042_domainaddress_used_on.py:9]: Nit: run `black` on this file

File: frontend/src/apiMocks/mockData.ts [hunk 2]
```diff
@@ -187,6 +187,7 @@ export const domainaddresses: Record<
       num_forwarded: 0,
       num_spam: 0,
       type: "custom",
+      used_on: "",
```
Comment (jwhitlock) [inline frontend/src/apiMocks/mockData.ts:190]: I get a `null` when this is unset for a fresh domain. Do we need to expect that on the React side, or enforce an empty string it on the Relay side?

**Review by jwhitlock — APPROVED** [review]
Thanks @groovecoder, the test worked locally for me. One nit and one question that @maxxcrawford may be able to answer.

File: frontend/src/apiMocks/mockData.ts (hunk 2 above)
Comment (Vinnl) [reply frontend/src/apiMocks/mockData.ts:190]: If it's supposed to be able to be `null` (i.e. the React side should expect that), we'll want to add ` | null` to the type in `frontend/src/hooks/api/aliases.ts`.

File: emails/migrations/0042_domainaddress_used_on.py (hunk 1 above)
Comment (groovecoder) [reply emails/migrations/0042_domainaddress_used_on.py:9]: /me shakes fist as double-quotes.

Comment (groovecoder) [reply emails/migrations/0042_domainaddress_used_on.py]: https://github.com/mozilla/fx-private-relay/pull/1968/commits/87f4b1c0d6783f96ea5c4986088085c4eae41d79

File: frontend/src/apiMocks/mockData.ts (hunk 2 above)
Comment (groovecoder) [reply frontend/src/apiMocks/mockData.ts:190]: I followed what I did on https://github.com/mozilla/fx-private-relay/pull/1713/files. Was that right? :)

Comment (groovecoder) [reply frontend/src/apiMocks/mockData.ts:190]: Updated type to have `| null` too.

File: api/serializers.py [hunk 3]
```diff
@@ -42,6 +44,7 @@ class Meta:
         ]
         read_only_fields = [
             "id",
+            "type",
```
Comment (maxxcrawford) [inline api/serializers.py:47]: issue: You're putting the `type` as read-only. This needs to be set when generating a mask from the add-on.

File: api/serializers.py [hunk 4]
```diff
@@ -55,12 +58,15 @@ class Meta:
 
 
 class DomainAddressSerializer(PremiumValidatorsMixin, serializers.ModelSerializer):
+    type = serializers.CharField(default="custom")
     class Meta:
         model = DomainAddress
         fields = [
+            "type",
```
Comment (maxxcrawford) [inline api/serializers.py:65]: suggestion: If this is read-only, it should go a few lines down, by "id" and "address".

**Review by maxxcrawford — CHANGES_REQUESTED** [review]

**Comment by groovecoder** [comment]
@jwhitlock do you mind checking this again? I made some changes to fix problems @maxxcrawford was having trying to use this.

**Review by maxxcrawford — APPROVED** [review]
Works as expected for me now. Thanks!

**Review by jwhitlock — APPROVED** [review]
Thanks @groovecoder! `used_on` and `mask_type` are now in the API for domain and random addresses:

domain

**Pull request merged**